File: src/main/java/net/minestom/server/utils/validate/Check.java
package net.minestom.server.utils.validate;

import org.jetbrains.annotations.Contract;
import org.jspecify.annotations.Nullable;

import java.text.MessageFormat;
import java.util.Objects;

/**
 * Convenient class to check for common exceptions.
 */
public final class Check {

    private Check() {

    }

    @Contract("null, _ -> fail")
    public static void notNull(@Nullable Object object, String reason) {
        if (Objects.isNull(object)) {
            throw new NullPointerException(reason);
        }
    }

    @Contract("null, _, _ -> fail")
    public static void notNull(@Nullable Object object, String reason, Object... arguments) {
        if (Objects.isNull(object)) {
            throw new NullPointerException(MessageFormat.format(reason, arguments));
        }
    }

    @Contract("true, _ -> fail")
    public static void argCondition(boolean condition, String reason) {
        if (condition) {
            throw new IllegalArgumentException(reason);
        }
    }

    @Contract("true, _, _ -> fail")
    public static void argCondition(boolean condition, String reason, Object... arguments) {
        if (condition) {
            throw new IllegalArgumentException(MessageFormat.format(reason, arguments));
        }
    }

    @Contract("_ -> fail")
    public static void fail(String reason) {
        throw new IllegalArgumentException(reason);
    }

    @Contract("_, _ -> fail")
    public static void fail(String reason, Object... arguments) {
        throw new IllegalArgumentException(MessageFormat.format(reason, arguments));
    }

    @Contract("true, _ -> fail")
    public static void stateCondition(boolean condition, String reason) {
        if (condition) {
            throw new IllegalStateException(reason);
        }
    }

    @Contract("true, _, _ -> fail")
    public static void stateCondition(boolean condition, String reason, Object... arguments) {
        if (condition) {
            throw new IllegalStateException(MessageFormat.format(reason, arguments));
        }
    }

    @Contract("false, _ -> fail")
    public static void isTrue(boolean condition, String reason) {
        if (!condition) {
            throw new IllegalStateException(reason);
        }
    }

    @Contract("false, _, _ -> fail")
    public static void isTrue(boolean condition, String reason, Object... arguments) {
        if (!condition) {
            throw new IllegalStateException(MessageFormat.format(reason, arguments));
        }
    }
}

File: src/main/java/net/minestom/server/utils/validate/Check.kt
package net.minestom.server.utils.validate

import org.jetbrains.annotations.Contract
import java.lang.NullPointerException
import java.text.MessageFormat
import java.lang.IllegalArgumentException
import java.lang.IllegalStateException
import java.util.*

/**
 * Convenient class to check for common exceptions.
 */
object Check {
    @Contract("null, _ -> fail")
    fun notNull(`object`: Any?, reason: String) {
        if (Objects.isNull(`object`)) {
            throw NullPointerException(reason)
        }
    }

    @Contract("null, _, _ -> fail")
    fun notNull(`object`: Any?, reason: String, vararg arguments: Any?) {
        if (Objects.isNull(`object`)) {
            throw NullPointerException(MessageFormat.format(reason, *arguments))
        }
    }

    @Contract("true, _ -> fail")
    fun argCondition(condition: Boolean, reason: String) {
        require(!condition) { reason }
    }

    @Contract("true, _, _ -> fail")
    fun argCondition(condition: Boolean, reason: String, vararg arguments: Any?) {
        require(!condition) { MessageFormat.format(reason, *arguments) }
    }

    @Contract("_ -> fail")
    fun fail(reason: String) {
        throw IllegalArgumentException(reason)
    }

    @Contract("true, _ -> fail")
    fun stateCondition(condition: Boolean, reason: String) {
        check(!condition) { reason }
    }

    @Contract("true, _, _ -> fail")
    fun stateCondition(condition: Boolean, reason: String, vararg arguments: Any?) {
        check(!condition) { MessageFormat.format(reason, *arguments) }
    }
}